Update docs for X-LoRA and some bugfixes

EricLBuehler opened the pull request:
Refs #1966. Adds some stuff to the docs, but the main focus is on the loading bugfix.

I ran:
```
ruff format
make quality
make style
pytest tests/test_xlora.py -v
```
With:
```
> ruff --version
ruff 0.5.7
> pytest --version
pytest 8.3.2
```

Looks like some tests are still failing, I'll take a look!

BenjaminBossan commented on examples/xlora_train/xlora_train_example_incomplete.ipynb:
```diff
@@ -0,0 +1,515 @@
+   },
+   "outputs": [],
+   "source": [
+    "FT_model_name = 'Zephyr_XLoRA' \n",
+    "output_dir=FT_model_name"
```
Can we get rid of one of those variables?

EricLBuehler replied:
Yep, both for now.

BenjaminBossan commented on examples/xlora_train/xlora_train_example_incomplete.ipynb:
```diff
@@ -0,0 +1,515 @@
+    "    xlora_dropout_p=0.2,\n",
+    ")\n",
+    "model = get_peft_model(model, peft_config).cuda()\n",
+    "print(model)\n"
```
My personal preference is not to print the models like this: It creates a lot of output that nobody reads and distracts from the important stuff IMO.

EricLBuehler replied:
Sounds good! I've removed it.

BenjaminBossan commented on examples/xlora_train/xlora_train_example_incomplete.ipynb:
```diff
@@ -0,0 +1,515 @@
+    "import numpy as np\n",
+    "device = torch.device(\"cuda\")\n",
+    "\n",
+    "token = 'hf_xxxxxxxxxxxxxxxxxxxx'\n",
+    "\n",
+    "from huggingface_hub import login\n",
+    "login(token=token)\n",
+    "\n",
```
Do we need the login code? If not, let's remove it.

EricLBuehler replied:
Yeah, not anymore!

BenjaminBossan commented on examples/xlora_train/xlora_train_example_incomplete.ipynb:
```diff
@@ -0,0 +1,515 @@
+   },
+   "outputs": [],
+   "source": [
+    "def params(model):\n",
```
This can be removed as it's not used AFAICT.

EricLBuehler replied:
Done!

BenjaminBossan commented on src/peft/tuners/xlora/model.py:
```diff
@@ -135,7 +200,40 @@ def __init__(
+            ephemeral_gpu_offload (`bool`, *optional*, defaults to `False`):
+                (For loading the LoRA adapters) Whether to use ephemeral GPU offloading for partially loaded modules.
+                Defaults to `False`.
+            autocast_adapter_dtype (`bool`, *optional*, defaults to `True`):
```
This argument is duplicated in the docstring.

EricLBuehler replied:
Thanks, fixed!

BenjaminBossan commented on src/peft/tuners/xlora/model.py:
```diff
@@ -86,6 +87,70 @@ def convert_layers_to_xlora(
+    """
+    All params pertain to the adapter (adapter name, model id, `i` is the adapter number in 0 indexing)
+    """
+    from peft.peft_model import PeftModel
```
Let's add a comment here that this "emulates" the functionality of `PeftModel.from_pretrained` and might need to be updated if there are changes there.

EricLBuehler replied:
Thanks, added that.

BenjaminBossan commented on src/peft/tuners/xlora/model.py:
```diff
@@ -86,6 +87,70 @@ def convert_layers_to_xlora(
+
+    # load the weights into the model
+    ignore_mismatched_sizes = kwargs.get("ignore_mismatched_sizes", False)
+    _load_result = set_peft_model_state_dict(
+        lora_model,
+        new_adapter_weights,
+        adapter_name=adapter_name,
+        ignore_mismatched_sizes=ignore_mismatched_sizes,
+    )
+    if len(_load_result.unexpected_keys) > 0:
```
No need for leading underscore

```suggestion
    load_result = set_peft_model_state_dict(
        lora_model,
        new_adapter_weights,
        adapter_name=adapter_name,
        ignore_mismatched_sizes=ignore_mismatched_sizes,
    )
    if len(load_result.unexpected_keys) > 0:
```

EricLBuehler replied:
I changes this, thanks!

BenjaminBossan commented on src/peft/tuners/xlora/model.py:
```diff
@@ -86,6 +87,70 @@ def convert_layers_to_xlora(
+    lora_model: LoraModel,
+    adapter_name: str,
+    model_id: str,
+    i: int,
```
`i` is not used, is it?

EricLBuehler replied:
Yep, fixed this!

BenjaminBossan reviewed (changes requested):
Thanks for the updates, nice that you could solve the loading issue.

Regarding the notebook: I added some comments but I would suggest to leave it out of the PR until it's finished. Also, WDYT about adding all the X-LoRA examples into a single directory `examples/xlora`? This may require updating some links.

It would be also great if we had add a test to ensure that the loading functionality works. For reference, this is what I used to test it (based on your notebook):

```python
import torch
from peft.tuners.xlora.config import XLoraConfig
from peft.utils.peft_types import PeftType, TaskType
from peft import get_peft_model
from safetensors.torch import load_file
from transformers import AutoModelForCausalLM

torch.manual_seed(123)

model_name= 'HuggingFaceH4/zephyr-7b-beta'
model = AutoModelForCausalLM.from_pretrained(model_name)
model.config.use_cache = False

adapters = [
    "lamm-mit/Zephyr_Bioinspired",
    "lamm-mit/Zephyr_CoT",
]
adapters = {str(i): file_name for i, file_name in enumerate(adapters)}

peft_config = XLoraConfig(
    task_type=TaskType.CAUSAL_LM,
    peft_type=PeftType.XLORA,
    hidden_size=model.config.hidden_size,
    adapters=adapters,
    xlora_depth=8,
    xlora_size=2048,
    layerwise_scalings=True,
    xlora_dropout_p=0.2,
)
model = get_peft_model(model, peft_config)

sd = load_file("examples/xlora_train/zephyr_bioinspired_adapter_model-fixed.safetensors")
w0 = model.base_model.model.model.layers[0].self_attn.q_proj.lora_A["0"].weight
w1 = sd['base_model.model.model.layers.0.self_attn.q_proj.lora_A.weight']

assert torch.allclose(w0, w1)
```

Of course, for a unit test, it should use a smaller model and ideally the adapter should be loaded from a HF-internal repo like https://huggingface.co/peft-internal-testing (LMK if you need access).

> I think the only CI failures are from spurious network errors

Yes, sometimes there can be timeouts, leading to failed tests. If you spot that, feel free to re-start the failing tests.

EricLBuehler commented on src/peft/tuners/xlora/classifier.py:
```diff
@@ -119,10 +119,14 @@ def make_dummy_scalings(
+                (batch_size, seq_len, self.n_layers, self.n_classes),
+                self.override_scaling_pass_value,
+            )
+            .to(device)
+            .to(self.dtype)
```
Thanks!

EricLBuehler commented:
> Regarding the notebook: I added some comments but I would suggest to leave it out of the PR until it's finished. Also, WDYT about adding all the X-LoRA examples into a single directory examples/xlora? This may require updating some links.

I have added this, I agree it makes more sense.

> It would be also great if we had add a test to ensure that the loading functionality works.
> Of course, for a unit test, it should use a smaller model and ideally the adapter should be loaded from a HF-internal repo like https://huggingface.co/peft-internal-testing (LMK if you need access).

That sounds great. I'll create a dummy adapter on a smaller model, as we only made X-LoRA Zephyr and Gemma 2 7b models. To upload the adapter, if you could grant access to https://huggingface.co/peft-internal-testing it would be great! Afterward, I can create the unit test.

BenjaminBossan commented:
LMK when this is ready for review.

EricLBuehler commented:
I think this should be ready for review! I don't plan any more changes.

BenjaminBossan reviewed (approved):
Thanks for these fixes and the improved docs. LGTM.

EricLBuehler commented:
Thanks!

benjamin-marie commented:
The code in the documentation (https://huggingface.co/docs/peft/main/en/package_reference/xlora) is deprecated.
It calls prepare_model_for_int8_training and passes "use_flash_attention_2" when loading the model.

EricLBuehler commented:
Hey @benjamin-marie! Thanks for pointing this out, I'll add a PR to fix this.
